File: binary_search.py
# 求[left, right) 内第一个不小于value的值的位置， 注意是左闭右开。
#example: 
#      array = [1,2,3] value = 4, left = 0, right = len(array) = 3
#      return 3

def lower_bound(array, left, right, value):
    while left < right:
        mid = left + (right - left) // 2
        if array[mid] < value:
            left = mid + 1
        else:
            right = mid
    return left

print(lower_bound([1,2,3], 0, 3, 2))

File: java/BinarySearch.java
import java.util.HashMap;

public class BinarySearch {
    public static int lowerBound(int[] nums, int value) {
        int left = 0;
        int right = nums.length;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == value) {
                right = mid;
            } else if (nums[mid] > value) {
                right = mid;
            } else if (nums[mid] < value) {
                left = mid + 1;
            }
        }
        return left;
    }

    public static int upperBound(int[] nums, int value) {
        int left = 0;
        int right = nums.length;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == value) {
                left = mid + 1;
            } else if (nums[mid] > value) {
                right = mid;
            } else if (nums[mid] < value) {
                left = mid + 1;
            }
        }
        return left;
    }

    public static void main(String args[]){
        int[] nums = new int[]{1, 2, 2, 3};
        System.out.println(lowerBound(nums, 2));
        System.out.println(upperBound(nums, 2));
        HashMap<Character, Integer> h1 = new HashMap<Character, Integer>();
    }

}
